File: src/main/java/tech/wintrich/networkapi/infrastructure/security/UrlValidator.java
package tech.wintrich.networkapi.infrastructure.security;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.net.InetAddress;
import java.net.URI;
import java.net.UnknownHostException;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Validator for URLs and hosts to prevent SSRF attacks
 */
@Slf4j
@Component
public class UrlValidator {

    @Value("${api.network.security.block-internal-ips:true}")
    private boolean blockInternalIps;

    @Value("#{'${api.network.security.blocked-hosts}'.split(',')}")
    private List<String> blockedHosts;

    private static final Pattern VALID_DOMAIN_PATTERN = 
            Pattern.compile("^([a-zA-Z0-9]([a-zA-Z0-9\\-]{0,61}[a-zA-Z0-9])?\\.)+[a-zA-Z]{2,}$");

    private static final Pattern VALID_IP_PATTERN = 
            Pattern.compile("^((25[0-5]|(2[0-4]|1\\d|[1-9]|)\\d)\\.?\\b){4}$");

    /**
     * Validates a URL for SSRF vulnerabilities
     *
     * @param url the URL to validate
     * @throws SecurityException if URL is invalid or blocked
     */
    public void validateUrl(String url) {
        try {
            URI uri = URI.create(url);
            String host = uri.getHost();

            if (host == null) {
                throw new SecurityException("Invalid URL: no host specified");
            }

            validateHost(host);

            // Only allow HTTP and HTTPS
            String scheme = uri.getScheme();
            if (scheme == null || (!scheme.equalsIgnoreCase("http") && !scheme.equalsIgnoreCase("https"))) {
                throw new SecurityException("Only HTTP and HTTPS protocols are allowed");
            }

        } catch (IllegalArgumentException e) {
            throw new SecurityException("Invalid URL format: " + e.getMessage());
        }
    }

    /**
     * Validates a host (domain or IP) for security issues
     *
     * @param host the host to validate
     * @throws SecurityException if host is invalid or blocked
     */
    public void validateHost(String host) {
        if (host == null || host.trim().isEmpty()) {
            throw new SecurityException("Host cannot be empty");
        }

        host = host.trim().toLowerCase();

        // Check blocked hosts list
        if (blockedHosts.contains(host)) {
            log.warn("Blocked host access attempt: {}", host);
            throw new SecurityException("Access to this host is not allowed");
        }

        // Check for localhost variants
        if (isLocalhost(host)) {
            log.warn("Localhost access attempt: {}", host);
            throw new SecurityException("Access to localhost is not allowed");
        }

        // Validate format (domain or IP)
        if (!isValidDomain(host) && !isValidIp(host)) {
            throw new SecurityException("Invalid host format");
        }

        // Check for internal IPs if blocking is enabled
        if (blockInternalIps) {
            try {
                InetAddress address = InetAddress.getByName(host);
                if (isInternalIp(address)) {
                    log.warn("Internal IP access attempt: {}", host);
                    throw new SecurityException("Access to internal IPs is not allowed");
                }
            } catch (UnknownHostException e) {
                // If we can't resolve it, let it pass - the actual service will handle it
                log.debug("Could not resolve host for IP check: {}", host);
            }
        }
    }

    /**
     * Checks if host is localhost
     */
    private boolean isLocalhost(String host) {
        return host.equals("localhost") ||
               host.equals("127.0.0.1") ||
               host.equals("0.0.0.0") ||
               host.equals("::1") ||
               host.startsWith("127.") ||
               host.equals("[::1]");
    }

    /**
     * Checks if the address is an internal/private IP
     */
    private boolean isInternalIp(InetAddress address) {
        return address.isLoopbackAddress() ||
               address.isLinkLocalAddress() ||
               address.isSiteLocalAddress() ||
               address.isAnyLocalAddress() ||
               isPrivateIpRange(address.getHostAddress());
    }

    /**
     * Checks if IP is in private ranges
     */
    private boolean isPrivateIpRange(String ip) {
        // 10.0.0.0/8
        if (ip.startsWith("10.")) return true;
        
        // 172.16.0.0/12
        if (ip.startsWith("172.")) {
            String[] parts = ip.split("\\.");
            if (parts.length >= 2) {
                int second = Integer.parseInt(parts[1]);
                if (second >= 16 && second <= 31) return true;
            }
        }
        
        // 192.168.0.0/16
        if (ip.startsWith("192.168.")) return true;
        
        // 169.254.0.0/16 (link-local)
        if (ip.startsWith("169.254.")) return true;
        
        return false;
    }

    /**
     * Validates domain format
     */
    private boolean isValidDomain(String domain) {
        return VALID_DOMAIN_PATTERN.matcher(domain).matches();
    }

    /**
     * Validates IP format
     */
    private boolean isValidIp(String ip) {
        return VALID_IP_PATTERN.matcher(ip).matches();
    }
}

File: web-api/Infrastructure/Security/UrlValidator.cs
using System.Net;
using System.Text.RegularExpressions;
using Wintrich.NetworkApi.Infrastructure.Exceptions;

namespace Wintrich.NetworkApi.Infrastructure.Security;

/// <summary>
/// Validates host names and URLs against SSRF attacks — direct port of the Java UrlValidator.
/// </summary>
public sealed class UrlValidator
{
    private readonly bool _blockInternalIps;
    private readonly HashSet<string> _blockedHosts;
    private readonly ILogger<UrlValidator> _logger;

    private static readonly Regex ValidDomainPattern =
        new(@"^([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}$",
            RegexOptions.Compiled, TimeSpan.FromSeconds(1));

    private static readonly Regex ValidIpPattern =
        new(@"^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}$",
            RegexOptions.Compiled, TimeSpan.FromSeconds(1));

    public UrlValidator(IConfiguration configuration, ILogger<UrlValidator> logger)
    {
        _logger = logger;
        _blockInternalIps = configuration.GetValue<bool>("Api:Network:Security:BlockInternalIps", true);
        var blocked = configuration.GetSection("Api:Network:Security:BlockedHosts").Get<string[]>()
                      ?? new[] { "localhost", "127.0.0.1", "0.0.0.0" };
        _blockedHosts = new HashSet<string>(blocked, StringComparer.OrdinalIgnoreCase);
    }

    /// <summary>Validates a full URL, checking scheme and delegating host validation.</summary>
    public void ValidateUrl(string url)
    {
        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri))
            throw new NetworkSecurityException("Invalid URL format");

        var scheme = uri.Scheme;
        if (!string.Equals(scheme, "http", StringComparison.OrdinalIgnoreCase) &&
            !string.Equals(scheme, "https", StringComparison.OrdinalIgnoreCase))
            throw new NetworkSecurityException("Only HTTP and HTTPS protocols are allowed");

        var host = uri.Host;
        if (string.IsNullOrWhiteSpace(host))
            throw new NetworkSecurityException("Invalid URL: no host specified");

        ValidateHost(host);
    }

    /// <summary>Validates a bare host or IP address against SSRF rules.</summary>
    public void ValidateHost(string host)
    {
        if (string.IsNullOrWhiteSpace(host))
            throw new NetworkSecurityException("Host cannot be empty");

        host = host.Trim().ToLowerInvariant();

        if (_blockedHosts.Contains(host))
        {
            _logger.LogWarning("Blocked host access attempt: {Host}", host);
            throw new NetworkSecurityException("Access to this host is not allowed");
        }

        if (IsLocalhost(host))
        {
            _logger.LogWarning("Localhost access attempt: {Host}", host);
            throw new NetworkSecurityException("Access to localhost is not allowed");
        }

        if (!IsValidDomain(host) && !IsValidIp(host))
            throw new NetworkSecurityException("Invalid host format");

        if (_blockInternalIps)
        {
            try
            {
                var addresses = Dns.GetHostAddresses(host);
                foreach (var address in addresses)
                {
                    if (IsInternalIp(address))
                    {
                        _logger.LogWarning("Internal IP access attempt: {Host} -> {Ip}", host, address);
                        throw new NetworkSecurityException("Access to internal IPs is not allowed");
                    }
                }
            }
            catch (NetworkSecurityException)
            {
                throw;
            }
            catch
            {
                // Cannot resolve — let the actual service handle it
                _logger.LogDebug("Could not resolve host for IP check: {Host}", host);
            }
        }
    }

    private static bool IsLocalhost(string host) =>
        host is "localhost" or "127.0.0.1" or "0.0.0.0" or "::1" or "[::1]"
        || host.StartsWith("127.", StringComparison.Ordinal);

    private static bool IsInternalIp(IPAddress address) =>
        IPAddress.IsLoopback(address)
        || address.IsIPv6LinkLocal
        || address.IsIPv6SiteLocal
        || IsPrivateRange(address);

    private static bool IsPrivateRange(IPAddress address)
    {
        var ip = address.ToString();

        // 10.0.0.0/8
        if (ip.StartsWith("10.", StringComparison.Ordinal)) return true;

        // 172.16.0.0/12
        if (ip.StartsWith("172.", StringComparison.Ordinal))
        {
            var parts = ip.Split('.');
            if (parts.Length >= 2 && int.TryParse(parts[1], out var second) && second >= 16 && second <= 31)
                return true;
        }

        // 192.168.0.0/16
        if (ip.StartsWith("192.168.", StringComparison.Ordinal)) return true;

        // 169.254.0.0/16 (link-local)
        if (ip.StartsWith("169.254.", StringComparison.Ordinal)) return true;

        return false;
    }

    private static bool IsValidDomain(string domain) => ValidDomainPattern.IsMatch(domain);
    private static bool IsValidIp(string ip) => ValidIpPattern.IsMatch(ip);
}
